fw/sock_clnt.c

```diff
@@ -620,28 +624,90 @@ tfw_sock_clnt_cfgend(void)
 	return 0;
 }
 
+static int
+tfw_listen_socks_array_cmp(const void *l, const void *r)
+{
+	TfwAddr *a = &(*(TfwListenSock **)l)->addr;
+	TfwAddr *b = &(*(TfwListenSock **)r)->addr;
+	int cmp = memcmp(&a->sin6_addr, &b->sin6_addr, sizeof(a->sin6_addr));
+
+	if (cmp)
+		return cmp;
+	else
+		return (a->sin6_port < b->sin6_port) ? -1 :
+	                                           (a->sin6_port > b->sin6_port);
+}
+
 /**
  * Start listening on all existing sockets (added via "listen" configuration
  * entries).
  */
 static int
 tfw_sock_clnt_start(void)
 {
-	int r;
+	int r = 0;
 	TfwListenSock *ls;
+	TfwListenSock *ls_reconf;
+	TfwListenSock *tmp;
+	size_t i, listen_socks_sz = tfw_listen_socks_sz;
+	TfwListenSock **listen_socks_array = kmalloc(tfw_listen_socks_sz *
+		                                         sizeof(listen_socks_array[0]),
+		                                         GFP_KERNEL);
+	TfwListenSock **ls_found;
+	bool *toched = kzalloc(tfw_listen_socks_sz, GFP_KERNEL);
+
+	i = 0;
+	list_for_each_entry(ls, &tfw_listen_socks, list)
+		listen_socks_array[i++] = ls;
+	sort(listen_socks_array, tfw_listen_socks_sz,
+		 sizeof(listen_socks_array[0]), tfw_listen_socks_array_cmp, NULL);
+
+	list_for_each_entry_safe(ls_reconf, tmp, &tfw_listen_socks_reconf, list) {
+		if (!(ls_found = bsearch(&ls_reconf, listen_socks_array,
+			                     tfw_listen_socks_sz,
+			                     sizeof(listen_socks_array[0]),
+			                     tfw_listen_socks_array_cmp))) {
```

Comment:
This gives `n*log(n) + n + m*log(n) = (n + m)*log(n) + n`, which is magnitude better than previous `n * m` - quite good!